### PR TITLE
Sanitize the optional dependencies for spider API

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/config/repository.py [hunk 1]
```diff
@@ -62,18 +70,28 @@ def read_config(self, config):
...
             config["thumbnail_sizes"]
         )
-        self.url_preview_enabled = config["url_preview_enabled"]
+        self.url_preview_enabled = config.get("url_preview_enabled", False)
         if self.url_preview_enabled:
             try:
+                import lxml
+                lxml  # To stop unused lint.
+            except ImportError:
+                raise ConfigError(MISSING_LXML)
```
Comment (ara4n) [inline synapse/config/repository.py]: Do these ConfigError exceptions actually get raised anywhere meaningful? When I tried it this way before, there was no logging set up whilst the config was being parsed, so they got swallowed.

File: synapse/python_dependencies.py [hunk 2]
```diff
@@ -43,7 +43,6 @@
         "matrix_angular_sdk>=0.6.8": ["syweb>=0.6.8"],
     },
     "preview_url": {
-        "lxml>=3.6.0": ["lxml"],
```
Comment (ara4n) [inline synapse/python_dependencies.py]: why remove lxml but keep netaddr?

File: synapse/rest/media/v1/preview_url_resource.py [hunk 3]
```diff
@@ -40,33 +40,11 @@
...
 
     def __init__(self, hs, filepaths):
-        try:
-            if html:
-                pass
-        except:
-            raise RuntimeError("Disabling PreviewUrlResource as lxml not available")
-
-        if not hasattr(hs.config, "url_preview_ip_range_blacklist"):
-            logger.warn(
```
Comment (ara4n) [inline synapse/rest/media/v1/preview_url_resource.py]: why have you removed the user-friendly log warning as to why the API was disabled?

File: synapse/python_dependencies.py (hunk 2 above)
Comment (erikjohnston) [inline synapse/python_dependencies.py]: Because netaddr is cheap to install with no external deps.

File: synapse/config/repository.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/config/repository.py]: https://github.com/matrix-org/synapse/blob/erikj/spider/synapse/app/homeserver.py#L288-L289

File: synapse/rest/media/v1/preview_url_resource.py (hunk 3 above)
Comment (erikjohnston) [inline synapse/rest/media/v1/preview_url_resource.py]: Woops. Added that back now. thanks.

File: synapse/config/repository.py [hunk 4]
```diff
@@ -13,10 +13,18 @@
...
 from collections import namedtuple
 
-import sys
+
+MISSING_NETADDR = (
+    "Missing netaddr library. This is required for URL preview API."
+)
+
+MISSING_LXML = (
+    "Missing lxml library. This is required for URL preview API."
```
Comment (NegativeMjark) [inline synapse/config/repository.py]: Maybe give ``pip install`` instructions for getting lxml?

File: synapse/python_dependencies.py (hunk 2 above)
Comment (NegativeMjark) [inline synapse/python_dependencies.py]: Do we say what version of lxml is needed anywhere?
@ara4n: What's the rational for requiring ``>=3.6.0`` ooi?

Comment (erikjohnston) [inline synapse/python_dependencies.py]: `lxml` doesn't come with a `__version__` fwiw.

Also, it seems to come as default in my python, (or at least its in the my system `dist-packages`)

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**